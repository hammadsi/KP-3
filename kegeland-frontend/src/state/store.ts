import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import patientsReducer from './ducks/patients/patients.reducer';
import questionnairesReducer from './ducks/questionnaires/questionnaires.reducer';
import { sensorReducer } from './ducks/sensors/sensors.reducer';
import sessionsReducer from './ducks/sessions/sessions.reducer';
import settingsReducer from './ducks/settings/settings.reducer';
import wheelchairPatientsReducer from './ducks/wheelchairPatients/wheelchairPatients.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  whitelist: ['settings', 'auth'],
};

export const rootReducer = combineReducers({
  settings: settingsReducer,
  auth: authReducer,
  patients: patientsReducer,
  sensors: sensorReducer,
  sessions: sessionsReducer,
  questionnaires: questionnairesReducer,
  wheelchairPatients: wheelchairPatientsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const isDev = !['production', 'test'].includes(process.env.NODE_ENV || '');

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(logger)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }),
  devTools: isDev,
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
