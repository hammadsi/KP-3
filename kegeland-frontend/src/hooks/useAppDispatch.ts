// eslint-disable-next-line @typescript-eslint/no-restricted-imports
import { useDispatch } from 'react-redux';

import { AppDispatch } from '../state/store';

/**
 * Custom hook for dispatching redux actions
 */
const useAppDispatch: () => AppDispatch = useDispatch;

export default useAppDispatch;
