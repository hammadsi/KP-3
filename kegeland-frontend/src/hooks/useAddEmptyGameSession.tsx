import { useState } from 'react';
import { addEmptyGameSession } from '../state/ducks/wheelchairPatients/wheelchairPatients.actions';
import useAppDispatch from './useAppDispatch';

const useAddEmptyGameSession = () => {
  const dispatch = useAppDispatch();
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  const addSession = async (patientId: string) => {
    setLoading(true);
    setError(null);
    try {
      await dispatch(addEmptyGameSession(patientId));
    } catch (e: any) {
      setError(`Failed to add empty game session: ${e.message}`);
    } finally {
      setLoading(false);
    }
  };

  return { addGameSession: addSession, loading, error };
};

export default useAddEmptyGameSession;
