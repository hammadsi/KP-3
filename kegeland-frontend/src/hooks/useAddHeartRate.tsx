import { useState } from 'react';

import { addHeartRateToGameSession } from '../state/ducks/wheelchairPatients/wheelchairPatients.actions';

import useAppDispatch from './useAppDispatch';

// This hook is for testing if the API call works. The actual call will be made from the Unity game.
const useAddHeartRate = () => {
  const dispatch = useAppDispatch();
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  const addHeartRate = async (
    patientId: string,
    sessionId: string,
    heartRate: number,
    timestamp: Date
  ) => {
    setLoading(true);
    setError(null);
    try {
      await dispatch(
        addHeartRateToGameSession({
          patientId,
          sessionId,
          heartRateData: { heartRate, timestamp }
        })
      );
    } catch (e: any) {
      setError(`Failed to add heart rate: ${e.message}`);
    } finally {
      setLoading(false);
    }
  };

  return { addHeartRate, loading, error };
};

export default useAddHeartRate;
