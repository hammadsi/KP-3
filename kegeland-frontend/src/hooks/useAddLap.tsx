import { useState } from 'react';
import { addLapToGameSession } from '../state/ducks/wheelchairPatients/wheelchairPatients.actions';
import useAppDispatch from './useAppDispatch';


const useAddLap = () => {
  const dispatch = useAppDispatch();
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  const addLap = async (
    patientId: string,
    sessionId: string,
    lapTime: number,
    timestamp: Date
  ) => {
    setLoading(true);
    setError(null);
    try {
      await dispatch(
        addLapToGameSession({
          patientId,
          sessionId,
          lapData: { lapTime, timestamp },
        })
      ).unwrap();  // Using unwrap to handle the returned promise correctly.
    } catch (e: any) {
      setError(`Failed to add lap data: ${e.message}`);
    } finally {
      setLoading(false);
    }
  };

  return { addLap, loading, error };
};

export default useAddLap;
