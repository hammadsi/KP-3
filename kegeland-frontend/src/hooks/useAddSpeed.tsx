import { useState } from 'react';

import { addSpeedToGameSession } from '../state/ducks/wheelchairPatients/wheelchairPatients.actions';
import useAppDispatch from './useAppDispatch';

const useAddSpeed = () => {
  const dispatch = useAppDispatch();
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<string | null>(null);

  const addSpeed = async (
    patientId: string,
    sessionId: string,
    leftSpeed: number,
    rightSpeed: number,
    timestamp: Date
  ) => {
    setLoading(true);
    setError(null);
    try {
      await dispatch(
        addSpeedToGameSession({
          patientId,
          sessionId,
          speedData: { leftSpeed, rightSpeed, timestamp }
        })
      );
    } catch (e: any) {
      setError(`Failed to add speed data: ${e.message}`);
    } finally {
      setLoading(false);
    }
  };

  return { addSpeed, loading, error };
};

export default useAddSpeed;
