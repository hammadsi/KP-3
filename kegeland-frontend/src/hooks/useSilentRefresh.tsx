import { useEffect, useState } from 'react';

import { refresh } from '../state/ducks/auth/auth.actions';

import useAppDispatch from './useAppDispatch';
import useAppSelector from './useAppSelector';

export const REFRESH_INTERVAL_MS = 50 * 60 * 1000;
const useSilentRefresh = () => {
  const dispatch = useAppDispatch();
  const { isSignedIn } = useAppSelector((state) => state.auth);
  const [timer, setTimer] = useState<NodeJS.Timer | null>(null);

  useEffect(() => {
    if (isSignedIn) {
      dispatch(refresh());
    }
  }, []);

  useEffect(() => {
    if (isSignedIn) {
      const interval = setInterval(() => {
        dispatch(refresh());
      }, REFRESH_INTERVAL_MS);
      setTimer(interval);
    } else {
      if (timer) {
        clearInterval(timer);
        setTimer(null);
      }
    }
    return () => {
      if (timer) {
        clearInterval(timer);
        setTimer(null);
      }
    };
  }, [isSignedIn]);

  return true;
};

export default useSilentRefresh;
