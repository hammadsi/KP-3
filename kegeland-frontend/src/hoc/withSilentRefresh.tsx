/* eslint-disable react/display-name */
import { Box } from '@chakra-ui/react';
import { ComponentProps, ComponentType } from 'react';

import useSilentRefresh from '../hooks/useSilentRefresh';

const withSilentRefresh =
  <P extends object>(
    Component: ComponentType<P>,
    compProps: ComponentProps<typeof Box>,
  ): ComponentType<P> =>
  (props) => {
    useSilentRefresh();
    return <Component {...props} {...compProps} />;
  };

export default withSilentRefresh;
