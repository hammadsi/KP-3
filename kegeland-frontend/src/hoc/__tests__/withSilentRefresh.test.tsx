import React from 'react';
import { Text } from '@chakra-ui/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import withSilentRefresh from '../withSilentRefresh';
import * as silentRefreshHook from '../../hooks/useSilentRefresh';
import { mockStore } from '../../state/mocks/store.mock';

describe('Test withAppWrapper-hoc', () => {
  const MockComponent: React.FC = () => {
    return <Text />;
  };

  it('should render correctly', () => {
    const Component = withSilentRefresh(MockComponent, {});
    const tree = renderer.create(
      <Provider store={mockStore()}>
        <MemoryRouter>
          <Component />
        </MemoryRouter>
      </Provider>,
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call correct hooks', () => {
    const silentRefreshHookSpy = jest.spyOn(silentRefreshHook, 'default');
    const Component = withSilentRefresh(MockComponent, {});
    renderer.create(
      <Provider store={mockStore()}>
        <MemoryRouter>
          <Component />
        </MemoryRouter>
      </Provider>,
    );
    expect(silentRefreshHookSpy).toBeCalled();
  });
});
