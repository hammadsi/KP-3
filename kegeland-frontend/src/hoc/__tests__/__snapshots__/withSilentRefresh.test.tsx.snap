// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test withAppWrapper-hoc should render correctly 1`] = `
<p
  className="chakra-text css-0"
/>
`;
